# **Notebook for Collection of Data**

## Objectives

* Fetch data from Kaggle and save as raw data in inputs/datasets/raw
* Inspect data fetched via Kaggle
* Save data fetched under outputs/datasets/collection

## Inputs

* My personal Kaggle JSON file to authenticate with Kaggle

## Outputs

* outputs/dataset/collection/house_price_records.csv
* outputs/dataset/collection/inherited_houses.csv
* Outputs were verified against the file names of the two csv files before Kaggle dowload was started
  
## Additional Comments

* File named inherited_houses.csv contains features and target for busines objective and is the basis for the ML model. The file named inherited_houses.csv contains features of houses that client would want to have predicted prices individually and as a total
 


---

# Change working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [ ]:
import os
current_dir = os.getcwd()
current_dir

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [ ]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

Confirm the new current directory

In [ ]:
current_dir = os.getcwd()
current_dir

# Collect data input from Kaggle

The Kaggle API was used to fetch the raw data for the project. For that purpose, the Kaggle package was installed.

The Kaggle API requires authentication with a token. A 'kaggle.json' is the token to authenticate with Kaggle.

The Kaggle token ('kaggle.json') was obtained as follows

- Log into the Kaggle acccount once created
- In the user profile picture select "Settings"
- A section named API is shown there
- Click "Expire API Token" to delete any previous tokens that may exist
- Click "Create New API Token". Once clicked, a new authentication token is created
- The authentication token created is a 'kaggle.json' file that must be downloaed
- The 'kaggle.json' file must be moved to the root directory of the project with an unchanged name. So, do not change the name 'kaggle.json' to a different file name or use a different extension other than .json


In [ ]:
%pip install kaggle==1.5.12

In [35]:
os.environ['KAGGLE_CONFIG_DIR'] = "C:/My_Folders/CodeInstitute/Project_5_files/Project-5"

The dataset used is located at the URL https://www.kaggle.com/datasets/codeinstitute/housing-prices-data. The Kaggle path and destination folder are defined as follows:

In [36]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "C:/My_Folders/CodeInstitute/Project_5_files/Project-5/inputs/datasets/raw"


The data is downloaded with the command in the subsequent cell. 

In [ ]:
! kaggle datasets download -d codeinstitute/housing-prices-data -p C:/My_Folders/CodeInstitute/Project_5_files/Project-5/inputs/datasets/raw


In [38]:
import zipfile
import os

zip_file_path = "C:/My_Folders/CodeInstitute/Project_5_files/Project-5/inputs/datasets/raw/housing-prices-data.zip"
destination_folder = "C:/My_Folders/CodeInstitute/Project_5_files/Project-5/inputs/datasets/raw"

if os.path.exists(zip_file_path):

    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
        zip_ref.extractall(destination_folder)
        

---

# Load and Review Data fetched from Kaggle

 - Load the Kaggle data into a pandas dataframe
 - Pandas library must be loaded first and subsequently dataset stored in pandas DataFrame

In [3]:
import pandas as pd

csv_file_path = r"C:\My_Folders\CodeInstitute\Project_5_files\Project-5\Project-5\inputs\datasets\raw\house-price-001\house_prices_records.csv"

# This line creates the dataframe for me base on the file_path defined above for the records of house prices in a csv format
df = pd.read_csv(csv_file_path)




- Make first few rows of data visible

In [4]:
df.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0        150            0.0         548          RFn  ...         65.0   
1        284            NaN         460          RFn  ...         80.0   
2        434            0.0         608          RFn  ...         68.0   
3        540            NaN         642          Unf  ...         60.0   
4        490            0.0         836          RFn  ...         84.0   

   MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  WoodDeckSF  \
0       196.0          61            5            7          856         0.0   
1         0.0           0            8            6         1262         NaN   
2       162.0          42            5            7          920         NaN   
3         0.0          35            5            7          756         NaN   
4       350.0          84            5            8         1145         NaN   

   YearBuilt  YearRemodAdd  SalePrice  
0       2003          2003     208500  
1       1976          1976     181500  
2       2001          2002     223500  
3       1915          1970     140000  
4       2000          2000     250000  

[5 rows x 24 columns]

- Obtain summary of the DataFrame

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1422 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1315 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1225 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

- Identify columns with missing values in the dataset

---

In [6]:
df.isnull().sum()

1stFlrSF            0
2ndFlrSF           86
BedroomAbvGr       99
BsmtExposure       38
BsmtFinSF1          0
BsmtFinType1      145
BsmtUnfSF           0
EnclosedPorch    1324
GarageArea          0
GarageFinish      235
GarageYrBlt        81
GrLivArea           0
KitchenQual         0
LotArea             0
LotFrontage       259
MasVnrArea          8
OpenPorchSF         0
OverallCond         0
OverallQual         0
TotalBsmtSF         0
WoodDeckSF       1305
YearBuilt           0
YearRemodAdd        0
SalePrice           0
dtype: int64

NOTE

- Check for duplicate rows in the dataset

In [7]:
df[df.duplicated(subset=None)]

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, EnclosedPorch, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, WoodDeckSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

[0 rows x 24 columns]

- Check the columns for unique values

In [8]:
for col in df:
    if df[col].dtypes == 'object':
        print(col, '-', df[col].unique())
    elif df[col].unique().size < 11:
        print(col, '-', df[col].unique().size)

BedroomAbvGr - 9
BsmtExposure - ['No' 'Gd' 'Mn' 'Av' nan]
BsmtFinType1 - ['GLQ' 'ALQ' 'Unf' 'Rec' nan 'BLQ' 'LwQ']
GarageFinish - ['RFn' 'Unf' nan 'Fin']
KitchenQual - ['Gd' 'TA' 'Ex' 'Fa']
OverallCond - 9
OverallQual - 10


Findings on the initial review of the dataset are

- the dataset consists of 1460 rows and 24 columns
- the data types include integers, floats, and objects
- no duplicate rows in the dataset
- 8 columns have missing values. For some columns almost all values are missing
- Columns such as BsmtExposure, BsmtFinType1 or GarageFinish contain categorical variables

Data cleaning and data preprocessing is needed to address missing values and treatment of categorical variables in the columns

---

# Push files to Repo

* Save the cleaned dataset to a local folder
* Push the dataset to the repository

In [9]:
import os
try:
  os.makedirs(name='outputs/datasets/collection')
except Exception as e:
  print(e)

df.to_csv(f"outputs/datasets/collection/house_prices.csv",index=False)


[WinError 183] Eine Datei kann nicht erstellt werden, wenn sie bereits vorhanden ist: 'outputs/datasets/collection'
